### PR TITLE
initialize dummy user auth

### LINKED ISSUE #21: initialize dummy user auth
**Todo:**
- create user auth middleware
- use jwt to transfer payload object from react to express
- setup dummy user auth

**Potential issues:**

**Possible solutions:**

**Additional notes (optional):**

### PR DESCRIPTION
**Pull request by 98sean98**
Closes #21 

**Important notes**
- a `jwt token` object should be sent by react to express to facilitate any user based operations
- `user` field in `booking` mongoose schema has been modified to `userId`, a `string` with `required: true`

### CODE REVIEW

File: express-app/middleware/authUser.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const jwt = require('jsonwebtoken');
+const to = require('await-to-js').default;
```
Comment (utkarsh867) [inline express-app/middleware/authUser.js]: why exactly do we need this?

File: express-app/middleware/authUser.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+const handleError = require('../routes/errorHandler');
+const key = process.env.API_KEY;
+const dummyUserId = process.env.DUMMY_USER_ID;
+
+const authUser = async (req, res, next) => {
+  // temporary jwt signature to verify payload
+  const token = req.body.token;
+  jwt.verify(token, key, (error, payload) => {
+    if (error || dummyUserId !== payload.userId)
```
Comment (utkarsh867) [inline express-app/middleware/authUser.js]: i think we can remove the dummyUserId check here

File: express-app/middleware/authUser.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const jwt = require('jsonwebtoken');
+const handleError = require('../routes/errorHandler');
+const key = process.env.API_KEY;
+
+const authUser = (req, res, next) => {
+  let token = req.headers['x-access-token'] || req.headers['authorization'];
```
Comment (utkarsh867) [inline express-app/middleware/authUser.js]: why the `||` operator?

File: express-app/routes/booking.js [hunk 4]
```diff
@@ -13,6 +13,10 @@ const Room = require('../models/room');
 // error handler
 const handleError = require('./errorHandler');
 
+// user auth middleware
+const authUser = require('../middleware/authUser.js');
+router.use('/create', authUser);
```
Comment (utkarsh867) [inline express-app/routes/booking.js:18]: why is the name of this route `/create`?

File: express-app/routes/booking.js [hunk 5]
```diff
@@ -46,22 +50,20 @@ const updateRoom = async (res, booking) => {
...
-    let savedRoom = await updateRoom(res, booking);
-    if (savedRoom.error) return;
-    let savedBooking = await createBooking(res, booking);
-    res.status(201).json({ savedBooking, savedRoom });
-    return;
-  }
-  handleError(res, 'Unauthorized access', 'Unauthorized access', 401);
+  const booking = req.body.booking;
+  booking.createdAt = new Date();
+  let newBooking = new Booking(booking);
```
Comment (utkarsh867) [inline express-app/routes/booking.js:58]: could probably be `const`

File: express-app/routes/booking.js [hunk 6]
```diff
@@ -46,22 +50,20 @@ const updateRoom = async (res, booking) => {
...
-      return handleError(res, 'Invalid booking', 'Invalid booking', 400);
-    let savedRoom = await updateRoom(res, booking);
-    if (savedRoom.error) return;
-    let savedBooking = await createBooking(res, booking);
-    res.status(201).json({ savedBooking, savedRoom });
-    return;
-  }
-  handleError(res, 'Unauthorized access', 'Unauthorized access', 401);
+  const booking = req.body.booking;
+  booking.createdAt = new Date();
```
Comment (utkarsh867) [inline express-app/routes/booking.js:57]: `createdAt` does not seem to be used at all in the mongoose model

File: express-app/routes/booking.js (hunk 5 above)
Comment (utkarsh867) [reply express-app/routes/booking.js:58]:
```
const booking = req.body.booking;
booking.createdAt = new Date();
if(newBooking.validateSync())
...
```

File: express-app/routes/booking.js [hunk 7]
```diff
@@ -46,22 +50,20 @@ const updateRoom = async (res, booking) => {
...
+  let newBooking = new Booking(booking);
+  let bookingError = await newBooking.validateSync();
+  if (bookingError)
+    return handleError(res, 'Invalid booking', 'Invalid booking', 400);
+  let savedRoom = await updateRoom(res, booking);
+  if (savedRoom.error) return;
+  let savedBooking = await createBooking(res, booking);
+  res.status(201).json({ savedBooking, savedRoom });
```
Comment (utkarsh867) [inline express-app/routes/booking.js]:
```
const savedRoom = ...
if (savedRoom.error) return;

const savedBooking = ...
...
```

File: express-app/routes/booking.js [hunk 8]
```diff
@@ -46,22 +50,20 @@ const updateRoom = async (res, booking) => {
...
+  return;
```
Comment (utkarsh867) [inline express-app/routes/booking.js]: redundant?

**Review by utkarsh867 — CHANGES_REQUESTED** [review]

File: express-app/middleware/authUser.js (hunk 1 above)
Comment (98sean98) [reply express-app/middleware/authUser.js]: `await-to-js` has been removed a long time ago

File: express-app/middleware/authUser.js (hunk 2 above)
Comment (98sean98) [reply express-app/middleware/authUser.js]: removed as well

File: express-app/middleware/authUser.js (hunk 3 above)
Comment (98sean98) [reply express-app/middleware/authUser.js]: should we use the `authorization` header only?

File: express-app/routes/booking.js (hunk 4 above)
Comment (98sean98) [reply express-app/routes/booking.js:18]: as far as the code base is being developed, `/booking/create` is the only route that requires user auth. so the middleware is passed to that, and nothing else

File: express-app/routes/booking.js (hunk 5 above)
Comment (98sean98) [reply express-app/routes/booking.js:58]: noted, will change

File: express-app/routes/booking.js (hunk 6 above)
Comment (98sean98) [reply express-app/routes/booking.js:57]: yes, it is. as a matter of fact, it is a `required: true` field in the `booking` model

File: express-app/routes/booking.js (hunk 7 above)
Comment (98sean98) [reply express-app/routes/booking.js]: noted, will change.

File: express-app/routes/booking.js (hunk 8 above)
Comment (98sean98) [reply express-app/routes/booking.js]: you refer to the last `return` statement? yes it appears so.
noted, will change.

**Comment by 98sean98** [comment]
Resolved requested changes. Please review again.

**Review by utkarsh867 — APPROVED** [review]

**Pull request merged**